server/sonar-server/src/main/java/org/sonar/server/usergroups/ws/SearchAction.java

```diff
@@ -0,0 +1,115 @@
+import org.sonar.api.server.ws.Request;
+import org.sonar.api.server.ws.Response;
+import org.sonar.api.server.ws.WebService.NewController;
+import org.sonar.api.server.ws.WebService.Param;
+import org.sonar.api.utils.text.JsonWriter;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.core.user.GroupDto;
+import org.sonar.server.db.DbClient;
+import org.sonar.server.es.SearchOptions;
+
+public class SearchAction implements UserGroupsWsAction {
+
+  public static final String FIELD_NAME = "name";
+  public static final String FIELD_DESCRIPTION = "description";
+  public static final String FIELD_MEMBERS_COUNT = "membersCount";
+
+  private DbClient dbClient;
+
+  public SearchAction(DbClient dbClient) {
+    this.dbClient = dbClient;
+  }
+
+  @Override
+  public void define(NewController context) {
+    context.createAction("search")
+      .setDescription("Search for user groups")
+      .setHandler(this)
+      .setResponseExample(getClass().getResource("example-search.json"))
+      .setSince("5.2")
+      .addFieldsParam(Arrays.asList(FIELD_NAME, FIELD_DESCRIPTION, FIELD_MEMBERS_COUNT))
+      .addPagingParams(100)
+      .addSearchQuery("sonar-users", "names");
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    int page = request.mandatoryParamAsInt(Param.PAGE);
+    int pageSize = request.mandatoryParamAsInt(Param.PAGE_SIZE);
+    SearchOptions options = new SearchOptions()
+      .setPage(page, pageSize);
+
+    String query = StringUtils.defaultIfBlank(request.param(Param.TEXT_QUERY), "");
+    List<String> fields = request.paramAsStrings(Param.FIELDS);
+
+    DbSession session = dbClient.openSession(false);
+    try {
+      int limit = dbClient.groupDao().countByQuery(session, query);
+      List<GroupDto> groups = dbClient.groupDao().selectByQuery(session, query, options.getOffset(), pageSize);
+      Collection<Long> groupIds = Collections2.transform(groups, new Function<GroupDto, Long>() {
+        @Override
+        public Long apply(@Nonnull GroupDto input) {
+          return input.getId();
+        }
+      });
+      Map<String, Integer> userCountByGroup = dbClient.groupMembershipDao().countUsersByGroups(session, groupIds);
+
+      JsonWriter json = response.newJsonWriter().beginObject();
+      options.writeJson(json, limit);
+      writeGroups(json, groups, userCountByGroup, fields);
+      json.endObject().close();
+    } finally {
+      session.close();
+    }
+  }
+
+  private void writeGroups(JsonWriter json, List<GroupDto> groups, Map<String, Integer> userCountByGroup, List<String> fields) {
+    json.name("groups").beginArray();
+    for (GroupDto group : groups) {
+      writeGroup(json, group, userCountByGroup.get(group.getName()), fields);
+    }
+    json.endArray();
+  }
+
+  private void writeGroup(JsonWriter json, GroupDto group, Integer memberCount, List<String> fields) {
+    json.beginObject()
+      .prop("key", group.getId())
```

Comment:
why not put the key field as a constant like the other fields ? I know it's used once, but at least we can have the list of fields at a glance